File: 남기석/redis/src/main/java/com/example/demo/infra/config/ExceptionHandlerFilter.java
package com.example.demo.infra.config;

import com.example.demo.infra.errors.ErrorCode;
import com.example.demo.infra.errors.ErrorResponse;
import com.example.demo.infra.errors.exception.LogOutedException;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@Component
@RequiredArgsConstructor
public class ExceptionHandlerFilter extends OncePerRequestFilter {
    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        try {
            filterChain.doFilter(request, response);
        } catch (LogOutedException e) {
            setErrorResponse(HttpStatus.BAD_REQUEST, response, e);
        }
    }

    private void setErrorResponse(HttpStatus status, HttpServletResponse response, LogOutedException e) {
        ErrorResponse errorResponse = new ErrorResponse(HttpStatus.BAD_REQUEST, ErrorCode.LOGOUT_ACCESS_TOKEN.getErrorCode(),
                ErrorCode.LOGOUT_ACCESS_TOKEN.getErrorMessage());
        try {
            response.setStatus(status.value());
            response.setContentType("application/json");
            response.getWriter().write(convertObjectToJson(errorResponse));
        } catch (IOException exception) {
            e.printStackTrace();
        }
    }

    public String convertObjectToJson(Object object) throws JsonProcessingException {
        if (object == null) {
            return null;
        }
        ObjectMapper mapper = new ObjectMapper();
        return mapper.writeValueAsString(object);
    }
}

File: Jose/redis/src/main/kotlin/com/jose/redis/exception/ExceptionHandlerFilter.kt
package com.jose.redis.exception

import com.fasterxml.jackson.databind.ObjectMapper
import com.jose.redis.exception.jwtException.HeaderHasNotAuthorization
import com.jose.redis.exception.jwtException.InvalidTokenException
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException
import java.time.LocalDateTime
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class ExceptionHandlerFilter(private val objectMapper: ObjectMapper) : OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        try {
            filterChain.doFilter(request, response)
        } catch (exception: InvalidTokenException) {
            return setErrorResponse(
                HttpStatus.BAD_REQUEST,
                request, response, exception.message!!, "토큰 오류 1번"
            )
        } catch (exception: HeaderHasNotAuthorization) {
            return setErrorResponse(
                HttpStatus.BAD_REQUEST,
                request, response, exception.message!!, "토큰 오류 2번"
            )
        }
    }

    fun setErrorResponse(status: HttpStatus, request: HttpServletRequest, response: HttpServletResponse, exceptionMessage: String, errorCode: String) {
        response.status = status.value()
        response.contentType = "application/json"
        val errorResponse =
            ErrorResponse(LocalDateTime.now(), HttpStatus.BAD_REQUEST.value(), errorCode, exceptionMessage, request.requestURI)
        try {
            response.characterEncoding = "UTF-8"
            response.writer.write(objectMapper.writeValueAsString(errorResponse))
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}
